[InstCombine] Pull extract through broadcast

agorenstein-nvidia opened the pull request:
The change adds a new instcombine pattern, and associated test, for patterns like this:

```
  %3 = shufflevector <2 x float> %1, <2 x float> poison, <4 x i32> zeroinitializer
  %4 = extractelement <4 x float> %3, i64 %idx
```

The shufflevector has a splat, or broadcast, mask, so the extractelement simply must be the first element of %1, so we transform this to

```
  %2 = extractelement <2 x float> %1, i64 0
```

agorenstein-nvidia committed:
Initial changes

agorenstein-nvidia committed:
Merge branch 'main' into instcombine-extract-through-broadcast

agorenstein-nvidia commented:
(Per github-actions bot, it seems I don't have the permissions to add reviewers directly. So...) Adding comment to request also my colleague @Prince781 to review this work. Thanks.

nikic reviewed (commented):
This seems like a generally profitable vector transform to me, but adding some more people to confirm.

agorenstein-nvidia commented on llvm/lib/Transforms/InstCombine/InstCombineVectorOps.cpp:
```diff
@@ -542,6 +542,12 @@ Instruction *InstCombinerImpl::visitExtractElementInst(ExtractElementInst &EI) {
     } else if (auto *SVI = dyn_cast<ShuffleVectorInst>(I)) {
+      // extractelt (shufflevector %v1, %v2, zeroinitializer) ->
+      // extractelt %v1, 0
+      if (isa<FixedVectorType>(SVI->getType()))
```
Thanks for taking such a quick look. To disambiguate, you're requesting I relax this constraint to include scalable vectors too? I'd be happy to do so and add the appropriate test.

nikic replied:
>  To disambiguate, you're requesting I relax this constraint to include scalable vectors too?

Yes, exactly.

RKSimon commented on llvm/lib/Transforms/InstCombine/InstCombineVectorOps.cpp:
```diff
@@ -542,6 +542,12 @@ Instruction *InstCombinerImpl::visitExtractElementInst(ExtractElementInst &EI) {
+      // extractelt (shufflevector %v1, %v2, zeroinitializer) ->
+      // extractelt %v1, 0
+      if (isa<FixedVectorType>(SVI->getType()))
+        if (all_of(SVI->getShuffleMask(), [](int Elt) { return Elt == 0; }))
```
This is probably worthwhile for any splat shuffle mask (assuming no undef mask elts)

agorenstein-nvidia committed:
PR feedback 1

agorenstein-nvidia committed:
clang-format

agorenstein-nvidia committed:
Fixing edit damage in test file

agorenstein-nvidia committed:
Removed unneeded check

agorenstein-nvidia committed:
Improve undef/poison usage in test (per PR automation failure), and using meaningful global instead of magic constant for shufflemask poison-indicator

agorenstein-nvidia committed:
Fixed varname convention per PR feedback; on reflection from PR feedback removing negative tests

agorenstein-nvidia commented on llvm/test/Transforms/InstCombine/vec_extract_through_broadcast.ll:
```diff
@@ -0,0 +1,54 @@
+  %3 = shufflevector <2 x float> %1, <2 x float> poison, <4 x i32> <i32 poison, i32 1, i32 1, i32 1>
+  %4 = extractelement <4 x float> %3, i64 %idx
+  ret float %4
+}
```
I see, I hadn't appreciated that, thanks for pointing it out. I am inclined to agree with the not-worth-bothering-with assessment.

However, learning this, I removed the two "negative" tests. While they helped start this conversation, I don't want to commit a test (inadvertently) enforcing that instcombine *doesn't* do a legal transformation, not when it seems like a reasonable one in principle (though, again, probably not worth bothering with!).

nikic replied:
Including the tests is generally a good idea, just should make it clear that this is just a missing optimization, not a negative test.

agorenstein-nvidia replied:
I believe I've appropriately integrated this feedback, so for book-keeping I'm resolving this thread. Please feel free to re-open if needed. Thanks for finding these ways to take this change further.

Prince781 reviewed (approved):
LGTM.

agorenstein-nvidia commented on llvm/lib/Transforms/InstCombine/InstCombineVectorOps.cpp:
```diff
@@ -542,6 +542,13 @@ Instruction *InstCombinerImpl::visitExtractElementInst(ExtractElementInst &EI) {
+      // extractelt (shufflevector %v1, %v2, splat-mask) idx ->
+      // extractelt %v1, splat-mask[0]
+      auto Mask = SVI->getShuffleMask();
+      if (Mask[0] != PoisonMaskElem && all_equal(Mask))
```
Thanks for catching that. With this and your other feedback, I'm seeing this transformation becoming more and more like the next case in the code (where the extract's constant is a known constant). That wasn't apparent to me at the beginning of the PR.

I'm exploring if there's a natural way to reframe this change as an extension to that prior work. I see in the docs that a mask's contents are either compile-time constants or poison, so using the just-found `llvm::getSplatIndex` (with care, it puns `not-found` and `poison`) we can basically reduce this new case to that the constant-index case. By that token, however, the constant case has some correctness concerns in the case of scalable vectors.

It is my EOD here, so I plan to return to this tomorrow with a refreshed look. Thanks again for all the quick and valuable feedback.

(I looked into the pipeline failures, the transformation kicked in more in related test files; I haven't closely examined the validity of those transformations yet, in light of the above.)

agorenstein-nvidia committed:
Restore old tests, reframe optimization as extension of existing one.

agorenstein-nvidia commented on llvm/test/Transforms/InstCombine/vscale_extractelement-inseltpoison.ll:
```diff
@@ -91,10 +91,7 @@ define i8 @extractelement_bitcast_insert_extra_use_bitcast(<vscale x 2 x i32> %a
-; CHECK-NEXT:    [[IN:%.*]] = insertelement <vscale x 4 x i32> poison, i32 [[V:%.*]], i64 0
-; CHECK-NEXT:    [[SPLAT:%.*]] = shufflevector <vscale x 4 x i32> [[IN]], <vscale x 4 x i32> poison, <vscale x 4 x i32> zeroinitializer
-; CHECK-NEXT:    [[R:%.*]] = extractelement <vscale x 4 x i32> [[SPLAT]], i64 4
-; CHECK-NEXT:    ret i32 [[R]]
```
This is analogous to the other comment; it *may* be that the `extractelement ... i32 4` is out of bounds (when `vscale=1`, so to speak). However, we are able to know that any valid index into `%splat` must be `%v`, so we just return `%v`. So, similar questions as the other comment.

RKSimon replied:
might be useful to add a comment above the test explaining the reasoning,similar to your comment here

agorenstein-nvidia commented:
Ping

agorenstein-nvidia committed:
fix formatting errors, sorry

agorenstein-nvidia commented:
> Ping

I noticed shortly after this ping someone (?) triggered some test actions, which revealed clang-format errors. I fixed them, and haven't seen any other issues. Now I believe I've addressed all outstanding comments/issues, and while I have questions about some code, they're of the sort "can this code be written better?" rather than change-of-behavior. For these reasons I believe this change is all-set, pending approvals---is there anything more I can do? Happy to take on as much as possible, please let me know.

RKSimon commented on llvm/test/Transforms/InstCombine/vec_extract_through_broadcast.ll:
```diff
@@ -0,0 +1,52 @@
+; CHECK-NEXT:    [[TMP1:%.*]] = extractelement <2 x float> [[W:%.*]], i64 0
+; CHECK-NEXT:    ret float [[TMP1]]
+;
+  %3 = shufflevector <2 x float> %1, <2 x float> poison, <4 x i32> zeroinitializer
```
(style) avoid using numbers for variable names (%vec, %shuffle and %extract would be better)

agorenstein-nvidia replied:
Done in latest push, please take a look. Thanks.

agorenstein-nvidia commented on llvm/lib/Transforms/InstCombine/InstCombineVectorOps.cpp:
```diff
@@ -542,27 +542,40 @@ Instruction *InstCombinerImpl::visitExtractElementInst(ExtractElementInst &EI) {
+        // getSplatIndex doesn't distinguish between the all-poison splat and
+        // a non-splat mask. However, if Index is -1, we still want to propagate
+        // that poison value.
+        int SrcIdx = -2;
```
Thanks for the suggestion and ping! I was out from Thursday until yesterday; this remains top of my "priority queue" and I'm integrating your suggestions now. Provisionally it looks like that exactly captures the intent, and clearly I hadn't considered this approach before. I'll finish putting this change in, and your test-improvement-suggestions (and run the tests/linter!) and push ASAP.

agorenstein-nvidia replied:
This has been done in the latest push, please take a look when you get the chance. It passes my local versions of git-clang-format and the lit tests, so hopefully that carries through to the CI here.

agorenstein-nvidia committed:
Improvement via optional from RKSimon's suggestion

agorenstein-nvidia committed:
Improved tests with semantic names per RKSimon's PR feedback

RKSimon reviewed (approved):
LGTM @nikic any more comments? whether we're happy with out of bounds extraction from a known splat (possibly with undef shuffle indices) seems to be the only outstanding issue

```
define i4 @src(i4 %x) {
  %v = insertelement <4 x i4> poison, i4 %x, i32 0
  %s = shufflevector <4 x i4> %v, <4 x i4> poison, 0, 0, 4294967295, 0
  %e = extractelement <4 x i4> %s, i32 4294967295
  ret i4 %e
}
=>
define i4 @tgt(i4 %x) {
  ret i4 %x
}
Transformation seems to be correct!
```

agorenstein-nvidia committed:
Improved comments to explicitly-named tests per PR suggestion

nikic reviewed (approved):
LGTM
